entrypoint.sh

```diff
@@ -26,36 +26,97 @@ echo "Starting: ACTION=$ACTION with DRY_RUN=$DRY_RUN"
-    if [[ $DRY_RUN == "True" ]]; then
-        $TERRAFORM_CMD plan -out=plan $TERRAFORM_VARS
-        if [ -f "validate_plan.py" ]; then
-            $TERRAFORM_CMD show -json $TERRAFORM_OUT_DIR/plan > $TERRAFORM_OUT_DIR/plan.json
-            python3 validate_plan.py "$TERRAFORM_OUT_DIR"/plan.json
-        fi
-    elif [[ $DRY_RUN == "False" ]]; then
-        $TERRAFORM_CMD apply -auto-approve $TERRAFORM_VARS
-        $TERRAFORM_CMD output -json > $OUTPUTS_FILE
-        if [ -f "post_checks.py" ]; then
-            python3 post_checks.py $OUTPUTS_FILE
+# the vars file path within the module directory
+export TERRAFORM_VARS="-var-file=tfvars.json"
+
+export PLAN_FILE="${APP}/plan"
+export PLAN_FILE_JSON="${APP}/plan.json"
+export OUTPUTS_FILE="${APP}/output.json"
+
+LOCK="-lock=true"
+if [[ $DRY_RUN == "True" ]]; then
+    LOCK="-lock=false"
+fi
+
+
+function run_hook() {
+    local HOOK_NAME="$1"
+    shift
+    local HOOK_DIR="./hooks"
+    local HOOK_SCRIPT=""
+
+    # Possible extensions for the hook scripts
+    local EXTENSIONS=("sh" "py")
+
+    if [ ! -d "$HOOK_DIR" ]; then
+        # no hook directory
+        return 0
+    fi
+
+    # Search for a valid hook script
+    for EXT in "${EXTENSIONS[@]}"; do
+        if [ -x "${HOOK_DIR}/${HOOK_NAME}.${EXT}" ]; then
+            HOOK_SCRIPT="${HOOK_DIR}/${HOOK_NAME}.${EXT}"
+            break
         fi
+    done
+
+    if [ -z "$HOOK_SCRIPT" ]; then
+        # no hook script
+        return 0
     fi
-elif [[ $ACTION == "Destroy" ]]; then
-    if [[ $DRY_RUN == "True" ]]; then
-        $TERRAFORM_CMD plan -destroy $TERRAFORM_VARS
-    elif [[ $DRY_RUN == "False" ]]; then
-        # Maybe the keys should be disabled instead of removed?
-        $TERRAFORM_CMD destroy -auto-approve $TERRAFORM_VARS
+
+    # Export variables for hooks
+    export DRY_RUN
+
+    echo "Running hook: $HOOK_NAME"
+    "$HOOK_SCRIPT" "$@"
+}
+
+function init() {
+    run_hook "pre_init"
+    $TERRAFORM_CMD init
+    run_hook "post_init"
+}
+
+
+function plan() {
+    run_hook "pre_plan"
+    PLAN_EXTRA_OPTIONS=""
+    if [[ $ACTION == "Destroy" ]]; then
+        PLAN_EXTRA_OPTIONS="-destroy"
     fi
-fi
+    $TERRAFORM_CMD plan ${PLAN_EXTRA_OPTIONS} -out=${PLAN_FILE} ${TERRAFORM_VARS} ${LOCK}
+    $TERRAFORM_CMD show -json ${PLAN_FILE} > ${PLAN_FILE_JSON}
+    run_hook "post_plan"
+}
+
+function apply() {
+    run_hook "pre_apply"
+    if [[ $ACTION == "Apply" ]] && [[ $DRY_RUN == "False" ]]; then
+        $TERRAFORM_CMD apply -auto-approve ${PLAN_FILE}
+        $TERRAFORM_CMD output -json > $OUTPUTS_FILE
+    elif [[ $ACTION == "Destroy" ]] && [[ $DRY_RUN == "False" ]]; then
+        $TERRAFORM_CMD destroy -auto-approve ${PLAN_FILE}
+    fi
+    run_hook "post_apply"
+}
+
+
+# Generate module configuration
+# generate-tf-config is a script in the final module.
+# It generates the terraform backend file and the vars file
+# into the module directory
+run_hook "pre_run"
+generate-tf-config
```

Comment:
`generate-tf-config` expect to be an executable in final module, right? Can we do a check to ensure it exists before run any hooks?

Reply:
Added a check.